### PR TITLE
Update color and icon previews

### LINKED ISSUE #3448: Color picker is not clickable in Android Studio
### Actual:
Clicking the color box, see picture, does nothing.
### Expected:
Clicking the color box should open the _**Color Picker**_ in _Android Studio_. 

bild

### LINKED ISSUE #3919: Color picker when specifying a color [Android Studio feature alike]
## Use case
Feeling a bit inconvenient when creating color widget in flutter, we need to manually insert the hexadecimal value for each color widget.

## Proposal
I think that will be great if we have color picker feature like in AS when we write hexadecimal value in our resource file. when we click the color preview there will be color picker pop up and let us pick the color we wanted. preview : [image](https://imgur.com/a/qkbAa0i)

### LINKED ISSUE #4986: Can't turn off color gutter icons
Color gutter icons stay visible, although it is turned off in settings. Only completly disable gutter icons hide the colors.

I tested with stable and EAP Intellij too. Neihter works.

IntelliJ IDEA 2020.3 EAP (Ultimate Edition)
Build #IU-203.4818.26, built on October 14, 2020
Flutter plugin 50.0

IntelliJ IDEA 2020.2.3 (Ultimate Edition)
Build #IU-202.7660.26, built on October 6, 2020
Flutter plugin 50.1.4

## Steps to Reproduce
Open project

## Version info

[✓] Flutter (Channel dev, 1.24.0-1.0.pre, on Mac OS X 10.15.7 19H2 darwin-x64, locale hu-HU)
    • Flutter version 1.24.0-1.0.pre at /Users/westito/flutter
    • Framework revision 12bea61c29 (6 days ago), 2020-10-14 15:57:04 -0700
    • Engine revision 069b3cf8f0
    • Dart version 2.11.0 (build 2.11.0-218.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/westito/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/westito/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Ultimate Edition (version 2020.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin installed
    • Dart plugin version 202.7319.5

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (4 available)
    • SM N950F (mobile) • ce0617165383e716027e • android-arm64  • Android 9 (API 28)
    • macOS (desktop)   • macos                • darwin-x64     • Mac OS X 10.15.7 19H2 darwin-x64
    • Web Server (web)  • web-server           • web-javascript • Flutter Tools
    • Chrome (web)      • chrome               • web-javascript • Google Chrome 86.0.4240.80

• No issues found!

### PR DESCRIPTION
**Pull request by stevemessick**
Change the implementation from annotation to line marker provider. This allows the gutter previews to be hidden when the system preference to do so is turned off.

This implementation is less rigid w.r.t formatting and comments, eg, this is recognized:
`Color . fromRGBO ( 0, 255 /* still green */, 0, 0.5 ) )`

It also adds a color picker when a color preview is clicked.
Screen Shot 2021-01-15 at 9 24 52 AM

Named colors, like `Color.blue`, also show the preview but cannot be changed with the color picker. The color picker will display, but the color will not be changed. Replacing non-leaf nodes in the PSI tree causes an error in the editor framework. We might be able to replace the text in the Document but we'd have to solve the problem of translating hex color values to named Material or Cupertino colors first.

The Dart plugin is updated to latest, which allows most unit tests to be run from IntelliJ. While debugging the problem with tests I realized a couple of testing helper classes should be declared abstract to make their role more clear.

Fixes #4986 
Fixes #3919
Fixes #3448

### CODE REVIEW

File: src/io/flutter/editor/FlutterColorProvider.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    }
+    else if (parent.getNode().getElementType() == DartTokenTypes.SIMPLE_TYPE) {
+      // const Color.fromARGB(100, 255, 0, 0)
+      // parent.getParent().getParent() is a new expr
+      parent = getNewExprFromType(parent);
+      if (parent == null) return null;
+      return parseColorElements(parent, refExpr);
+    }
+    else {
+      // name.equals(refExpr.getFirstChild().getText()) -> Colors.blue
```
Comment (helin24) [inline src/io/flutter/editor/FlutterColorProvider.java:65]: I'm not sure what this comment and the next one express - is this condition more complex than when the color is presented in the form `Colors.blue`?

**Review by helin24 — APPROVED** [review]
LGTM, and the tests are helpful

File: src/io/flutter/editor/FlutterColorProvider.java (hunk 1 above)
Comment (stevemessick) [reply src/io/flutter/editor/FlutterColorProvider.java:65]: These are "notes to self" (or next maintainer) about the structure of the PSI tree. JetBrains' Dart parser produces an AST in a style I remember reading about in my compilers class, but never actually used in any compiler I worked on. I work in the PSI tree every couple years, it seems, and it takes me some time to get back up to speed. These notes should help for next time.

In this case `name` is "Colors" and the first child of the reference expression is also "Colors" so we know we're looking at `Colors.blue` or some other named color. The next case is the structure for a shade reference, like `Colors.blue.shade200`.

**Comment by devoncarew** [comment]
A comment on this PR:

Note that the previous icon and color code had a performance issue we found after the fact. We were performing some expensive operations while checking the PSI tree elements. It was easy to correct after it had been identified via profiling. It might be worth profiling this new code on large files to validate that there aren't any unexpected perf changes.

**Comment by stevemessick** [comment]
Added #5188

**Pull request merged**